### ISSUE TITLE
HTTPError 403 fetching web pages

### ISSUE THREAD

**Issue by Ryo-Sajima** [cinemagoer/cinemagoer#415]
#### Issue description
As in the description.

#### Version of Cinemagoer, Python and OS
- **Python:** Python 3.9.2
- **Cinemagoer:** 2022.02.11
- **OS:** uname_result(system='Linux', node='raspberrypi', release='5.15.61+', version='#1579 Fri Aug 26 11:08:59 BST 2022', machine='armv6l')

#### Steps to reproduce the issue

```python
import imdb
ia = imdb.Cinemagoer()
ia.search_movie('The Curious Case of Benjamin Button')
```
Try multiple times. Occurs only sometimes.

#### What's the expected result?
```
[<Movie id:0421715[http] title:_The Curious Case of Benjamin Button (2008)_>, <Movie id:3011584[http] title:_"Zoom In" The Curious Case of Benjamin Button (2008)_>, <Movie id:7229412[http] title:_"Chris Stuckmann Movie Reviews" The Curious Case of Benjamin Button (2014)_>, <Movie id:18793490[http] title:_The Curious Case of Benjamin Button by F. Scott Fitzgerald (Short Story Month) (2019) (Podcast Episode) - The Drunk Guys Book Club Podcast (2017)_>, <Movie id:17184240[http] title:_The Curious Case of Benjamin Button by F. Scott Fitzgerald (2021) (Podcast Episode) - Just Sleep - Bedtime Stories for Adults (2021)_>, <Movie id:19831222[http] title:_The Curious Case of Benjamin Button âEUR¢ The Next Reel (2012) (Podcast Episode) - The Next Reel Film Podcast Master Feed (2011)_>, <Movie id:3487928[http] title:_"Partners" The Curious Case of Benjamin Butt-Ugly (2014)_>, <Movie id:1430122[http] title:_The Curious Birth of Benjamin Button (2009) (V)_>, <Movie id:8343064[http] title:_"RealTVFilms" Alexandre Desplat, Curious Case of Benjamin Button (2009)_>, <Movie id:22311096[http] title:_The Curious Case of Benja-Scam Button w/ Josh Brekhus (2022) (Podcast Episode) - Scam Goddess (2019)_>, <Movie id:17497930[http] title:_"SSDM" What is THE CURIOUS CASE OF BENJAMIN BUTTON really about? (2021)_>, <Movie id:2135211[http] title:_"Hollywood on Set" Ink Heart/Paul Blart: Mall Cop/The Curious Case of Benjamin Button (2009)_>, <Movie id:18151852[http] title:_Bavarian Podcast Works Flagship Show S4E14: Rotations for Bayern, the futures for Sabitzer & Musiala, the curious case of Benjamin Pavard, the Ballon d'Or, and more! (2021) (Podcast Episode) - Bavarian Football Works: For Bayern Munich fans (2019)_>]
```

#### What's the actual result?
```
Traceback (most recent call last):
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 221, in retrieve_unicode
    response = uopener.open(url)
  File "/usr/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/usr/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 405: Not Allowed
Traceback (most recent call last):
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 221, in retrieve_unicode
    response = uopener.open(url)
  File "/usr/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/usr/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 405: Not Allowed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/__init__.py", line 444, in search_movie
    res = self._search_movie(title, results)
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 408, in _search_movie
    cont = self._get_search_content('tt', title, results)
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 396, in _get_search_content
    cont = self._retrieve(self.urls['find'] % params)
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 379, in _retrieve
    ret = self.urlOpener.retrieve_unicode(url, size=size)
  File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 249, in retrieve_unicode
    raise IMDbDataAccessError(
imdb._exceptions.IMDbDataAccessError: {'errcode': None, 'errmsg': 'None', 'url': 'https://www.imdb.com/find?q=The+Curious+Case+of+Benjamin+Button&s=tt', 'proxy': '', 'exception type': 'IOError', 'original exception': <HTTPError 405: 'Not Allowed'>}
```

**Comment by WereCatf** [comment]
It appears IMDb doesn't allow requests that don't specify a user-agent anymore. Simply adding e.g. the following line to imdb/parser/http/__init__.py at line 219 fixed the issue for me:

`self.set_header("User-Agent", "alberanid/Cinemagoer")`

**Comment by phoenixwu** [comment]
> It appears IMDb doesn't allow requests that don't specify a user-agent anymore. Simply adding e.g. the following line to imdb/parser/http/**init**.py at line 219 fixed the issue for me:
> 
> `self.set_header("User-Agent", "alberanid/Cinemagoer")`

I still got 403 forbidden error with this User-Agent.
but this solution WORKS perfectly after I changed a longer User-Agent, such as "Mozilla/5.0 (X11; Linux i686; rv:107.0) Gecko/20100101 Firefox/107.0"

thanx

**Comment by Ryo-Sajima** [comment]
I changed the User-Agent as
```python
self.set_header('User-Agent', 'Mozilla/5.0 (X11; CrOS armv6l 13597.84.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36 Edg/107.0.1418.56')
```
and now, the error does not pop up, but the result is sometimes not correct.

Tried this code, with the user-agent above:
```python
import imdb
ia = imdb.Cinemagoer()
for _ in range(10):
    print(len(ia.search_movie('The Curious Case of Benjamin Button')))
```

and got this:
```
13
13
0
13
0
0
13
13
0
0
```

**Comment by alberanid** [comment]
@Ryo-Sajima thanks, I've updated the User-Agent header. Let's see if it's enough.

**Comment by meeotch** [comment]
I ran the test script from #418 and got a syntax error - looks like there's an extra quote and parenthesis:

```
Traceback (most recent call last):
  File "./test_imdb.py", line 13, in <module>
    ia = Cinemagoer()
  File "/home/mitch/.local/lib/python3.6/site-packages/imdb/__init__.py", line 195, in IMDb
    from .parser.http import IMDbHTTPAccessSystem
  File "/home/mitch/.local/lib/python3.6/site-packages/imdb/parser/http/__init__.py", line 165
    'Mozilla/5.0 (X11; CrOS armv6l 13597.84.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36 Edg/107.0.1418.56')')
```
I fixed the error in parser/html/__init__.py and got a similar result as before, though less inconsistent:  get_person() failed the first time, but succeeded the second.  Subsequent runs of the test script worked on the first try, but then leaving it idle for a couple minutes and running it again failed on the first time through the loop, and succeeded on the second.

So overall, much better, if not totally fixed.  If you're testing, you may have to avoid making requests for a few minutes in between runs.

```
['headshot', 'birth name', 'height', 'mini biography', 'trivia', 'quotes', 'salary history', 'birth date', 'birth notes', 'full-size headshot']
Trying person 0006904 again...
['birth info', 'headshot', 'filmography', 'imdbID', 'name', 'imdbIndex', 'birth name', 'height', 'mini biography', 'trivia', 'quotes', 'salary history', 'birth date', 'birth notes', 'canonical name', 'long imdb name', 'long imdb canonical name', 'full-size headshot']
```

**Comment by alberanid** [comment]
@meeotch thanks for the notice! My fault, I've pushed a fix for the syntax error.

We will see if it's a temporary condition or if there's something else going on.

**Comment by uyar** [comment]
When working on the test suite, I noticed that the problems occur on the "find" pages (see #411) and interestingly the person main page but not on any other page (not yet at least). I've added a delay setting to the test suite to see if adding some delay between requests would resolve things but didn't see any consistent behavior.

**Comment by alberanid** [comment]
Closing this issue since right now the test suite works fine. If it happens again, please reopen the issue (or open a new one). Thanks!

**Comment by Ryo-Sajima** [comment]
I'm now getting a different error.

```
2022-12-21 14:15:37,625 CRITICAL [imdbpy] C:\Users\Roy\anaconda3\lib\site-packages\imdb\_exceptions.py:32: IMDbDataAccessError exception raised; args: ({'errcode': None, 'errmsg': 'None', 'url': 'https://www.imdb.com/find?q=The+Curious+Case+of+Benjamin+Button&s=tt', 'proxy': '', 'exception type': 'IOError', 'original exception': <HTTPError 308: 'Permanent Redirect'>},); kwds: {}
Traceback (most recent call last):
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\parser\http\__init__.py", line 222, in retrieve_unicode
    response = uopener.open(url)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 523, in open
    response = meth(req, response)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 632, in http_response
    response = self.parent.error(
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 561, in error
    return self._call_chain(*args)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 494, in _call_chain
    result = func(*args)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 308: Permanent Redirect
Traceback (most recent call last):
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\parser\http\__init__.py", line 222, in retrieve_unicode
    response = uopener.open(url)
    response = meth(req, response)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 632, in http_response
    response = self.parent.error(
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 561, in error
    return self._call_chain(*args)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 494, in _call_chain
    result = func(*args)
  File "C:\Users\Roy\anaconda3\lib\urllib\request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 308: Permanent Redirect

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\__init__.py", line 450, in search_movie
    res = self._search_movie(title, results)
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\parser\http\__init__.py", line 410, in _search_movie
    cont = self._get_search_content('tt', title, results)
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\parser\http\__init__.py", line 398, in _get_search_content
    cont = self._retrieve(self.urls['find'] % params)
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\parser\http\__init__.py", line 381, in _retrieve
    ret = self.urlOpener.retrieve_unicode(url, size=size)
  File "C:\Users\Roy\anaconda3\lib\site-packages\imdb\parser\http\__init__.py", line 250, in retrieve_unicode
    raise IMDbDataAccessError(
imdb._exceptions.IMDbDataAccessError: {'errcode': None, 'errmsg': 'None', 'url': 'https://www.imdb.com/find?q=The+Curious+Case+of+Benjamin+Button&s=tt', 'proxy': '', 'exception type': 'IOError', 'original exception': <HTTPError 308: 'Permanent Redirect'>}
```

**Comment by xalienxx** [comment]
> I'm now getting a different error.
> 
> ```
> 2022-12-21 14:17:13,274 CRITICAL [imdbpy] /home/roy/.local/lib/python3.9/site-packages/imdb/_exceptions.py:32: IMDbDataAccessError exception raised; args: ({'errcode': None, 'errmsg': 'None', 'url': 'https://www.imdb.com/find?q=The+Curious+Case+of+Benjamin+Button&s=tt', 'proxy': '', 'exception type': 'IOError', 'original exception': <HTTPError 308: 'Permanent Redirect'>},); kwds: {}
> Traceback (most recent call last):
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 222, in retrieve_unicode
>     response = uopener.open(url)
>   File "/usr/lib/python3.9/urllib/request.py", line 523, in open
>     response = meth(req, response)
>   File "/usr/lib/python3.9/urllib/request.py", line 632, in http_response
>     response = self.parent.error(
>   File "/usr/lib/python3.9/urllib/request.py", line 561, in error
>     return self._call_chain(*args)
>   File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.9/urllib/request.py", line 641, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> urllib.error.HTTPError: HTTP Error 308: Permanent Redirect
> Traceback (most recent call last):
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 222, in retrieve_unicode
>     response = uopener.open(url)
>   File "/usr/lib/python3.9/urllib/request.py", line 523, in open
>     response = meth(req, response)
>   File "/usr/lib/python3.9/urllib/request.py", line 632, in http_response
>     response = self.parent.error(
>   File "/usr/lib/python3.9/urllib/request.py", line 561, in error
>     return self._call_chain(*args)
>   File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.9/urllib/request.py", line 641, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> urllib.error.HTTPError: HTTP Error 308: Permanent Redirect
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/__init__.py", line 450, in search_movie
>     res = self._search_movie(title, results)
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 410, in _search_movie
>     cont = self._get_search_content('tt', title, results)
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 398, in _get_search_content
>     cont = self._retrieve(self.urls['find'] % params)
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 381, in _retrieve
>     ret = self.urlOpener.retrieve_unicode(url, size=size)
>   File "/home/roy/.local/lib/python3.9/site-packages/imdb/parser/http/__init__.py", line 250, in retrieve_unicode
>     raise IMDbDataAccessError(
> imdb._exceptions.IMDbDataAccessError: {'errcode': None, 'errmsg': 'None', 'url': 'https://www.imdb.com/find?q=The+Curious+Case+of+Benjamin+Button&s=tt', 'proxy': '', 'exception type': 'IOError', 'original exception': <HTTPError 308: 'Permanent Redirect'>}
> ```

IMDB has changed the URL structure 
from :
https://www.imdb.com/find?q=the+batman&s=tt
to :
https://www.imdb.com/find/?q=the%20batman&s=tt

I think the difference is that " /find?q= " has been used instead of " /find/?q= "

**Comment by alberanid** [comment]
@Ryo-Sajima fixed in #426 ; you can use the version in the repository for the moment. I will release a new version on pypi in the next days.
